src/commands/command_parser.h

```diff
@@ -56,12 +60,39 @@ struct CommandParser {
 
   decltype(auto) RawPeek() const { return *begin_; }
 
+  decltype(auto) operator[](size_t index) const { return begin_[index]; }
+
   decltype(auto) RawTake() { return *begin_++; }
 
   decltype(auto) RawNext() { ++begin_; }
 
   bool Good() const { return begin_ != end_; }
 
+  std::enable_if_t<IsRandomAccessIter(), size_t> Remains() const {
+    // O(1) iff Iter is random access iterator.
+    auto d = std::distance(begin_, end_);
+    if (d < 0) {
+      d = 0;
+    }
```

Comment:
I think maybe we can just DCHECK or assert here, then we can find problem when we encounter an invalid state.